### ISSUE TITLE
An unusually large amount of memory usage

### ISSUE THREAD

**Issue by pflannery** [bevry/taskgroup#19]
When creating 10000 small tasks I find that the process memory peeks around 170MB. When I run the same 10000 tasks with setImmediate it's doesn't go over 40MB.
So its seems that task group is generating an unusually large amount of memory overhead.

Here the code I was testing with.
```js
var TaskGroup = require("taskgroup").TaskGroup;
var testArray = Array(10000).join().split(',');
var grp = TaskGroup.create();

testArray.forEach(function(value, index) {
  grp.addTask(function() {
    console.log("task"+index);
  });
  /*
  setImmediate(function() {
    console.log("task"+index);
  });
  */
});

grp.run();
```

I only tested on Windows 8.1 so not sure what a Linux box will output

**Comment by balupton** [comment]
Perhaps linked to https://github.com/bevry/taskgroup/issues/12

**Comment by balupton** [comment]
To be honest, I would open to migrating TaskGroup to a pre-processor that actually cares about performance. CoffeeScript no longer cares about it at all, which is really sad. Perhaps CoffeeScript Redux is an option, otherwise we could look at TypeScript, or maybe just native javascript (not that it is actually a good language - unlike compared to 5 years ago)

**Comment by pflannery** [comment]
Seeing as the topic has gone and the ssl is telling me that it's not valid. I will just reply here for now :)

I agree, I think moving to native ES5\ES6 will give us more control over what code we use. I'm all for it.
I find coffee classes are a real pain when analysing them in a heap snapshot because they are all called ```ctor``` instead of the class names I define. This makes it very difficult.

I've been using v8-profiler to take snapshots of the test code I posted above and below you see that a Task instance is around 8K and the _events property generated by the EventEmitter is adding a 5-6kb per Task. So I'm wonder if its worth replacing the event emitter for something that is much lighter in weight? 

![image](https://cloud.githubusercontent.com/assets/1727302/6763498/837700aa-cf77-11e4-9606-90cf480dd36d.png)

**Comment by balupton** [comment]
https://discuss.bevry.me/t/move-from-coffeescript-to-es6/30 is the new link :-)

**Comment by pflannery** [comment]
Just want to correct myself slightly, EventEmitter isn't adding 5-6k per task, instead it just seems that its referencing 5-6k.

**Comment by balupton** [comment]
@pflannery can you do a screencast sometime of how you come up with that profile information - I've spent ages trying to figure it out and no luck!

**Comment by pflannery** [comment]
Here's the code for generating the heap snapshot..

```js
var fs = require('fs');
var profiler = require('v8-profiler');

function saveSnapshot(fileName) {
  var buffer = '';
  var snapshot = profiler.takeSnapshot("test");
  var t = snapshot.serialize(
    function iterator(data, length) {
      buffer += data;
    },function complete(){
      fs.writeFileSync(fileName, buffer)
    }
  );
}

var TaskGroup = require("./taskgroup").TaskGroup;
var testArray = Array(10000).join().split(',');
var grp = TaskGroup.create();

testArray.forEach(function(value, index) {
  grp.addTask(function() {
    // console.log("task"+index);
  }); 
});

grp.on("completed", function(){
  saveSnapshot("completed.heapsnapshot")
});

grp.run();
```

**Comment by pflannery** [comment]
also here is a Memory Management Masterclass video on https://www.youtube.com/watch?v=LaxbdIyBkL0

**Comment by pflannery** [comment]
I've created a gist with javascript that generates the cpu profile and ensures the deopt reasons will render in latest chrome - https://gist.github.com/pflannery/8e38a06a844a7fc362dc

##### setup instructions
- create a temp folder
- ```npm install taskgroup```
- ```npm install v8-profiler```
- run ```node taskgroup-cpuprofiling.js``` 
- open the generated ```complete.cpuprofile``` file in ```chrome-dev-tools -> profiler```

**Comment by balupton** [comment]
Sweet. I'll give it a go (watching the talk right now). I'm wondering if it is best to just do the performance tracking inside the browser with a browserified/webpack'd build of taskgroup. Your thoughts?

**Comment by pflannery** [comment]
I've never used browserified or webpack so don't know what the outcome would be.

**Comment by pflannery** [comment]
@balupton It wouldn;t let me post post than three times on the [discuss thread](https://discuss.bevry.me/t/move-from-coffeescript-to-es6/30/11) and it wont let me add any more links in the post either so I've replied here

I've been analysing the heap generated from running the my [test](https://github.com/bevry/taskgroup/issues/19#issue-62773051) and i've seen that commenting out the following lines takes the heap snapshot from 114MB down to 31MB. (the snapshot was taken once during the taskgroup completed event)

[Lines here](https://github.com/bevry/taskgroup/blob/es6/es6/lib/taskgroup.js#L336-L340) and [Lines here](https://github.com/bevry/taskgroup/blob/es6/es6/lib/taskgroup.js#L344-L351)

So from that I think this confirms that the EventEmitter (taking up 53MB) is a big part of this problem and it also seems that the Map copying (taking up 14MB) is adding to it too.

**Comment by balupton** [comment]
Interesting. A few notes:

Commenting out those events means that the taskgroup of 31MB doesn't work...

Perhaps the downsize of the commented out map options are actually because it disables domains.

**Comment by balupton** [comment]
@pflannery think you came across these rate limits: https://www.dropbox.com/s/sltz6p9gavq64vz/Screenshot%202015-03-24%2008.25.01.png?dl=0

Tomorrow it won't be a problem :-)

**Comment by balupton** [comment]
@pflannery I've pushed some profiling things up to the `es6` branch, you can now run:

```
cake compile
npm run-script profile
```

To generate the V8 profile report.

However, you can also run:

```
cake compile
npm run-script browserify
```

Then serve the `web` to your browser, and then profile from there! Got Chrome and Firefox both profiling :+1:

**Comment by balupton** [comment]
Here's the profile that chrome generated: https://www.dropbox.com/s/alsdwj0u97sm09c/taskgroup-es6-babel.cpuprofile?dl=0

**Comment by balupton** [comment]
Conclusions from evaluating the running of tasks

1. A lot of time is spent emitting events, as such we should avoid emitting unnecessary events (this is what you found by commenting out the event names)
1. A lot of time is spent waiting between tasks due to the setImmediate/nextTick calls to prevent locking, as such we should look into alternative ways of queuing tasks without causing lockups
1. Creating a domain in the browser (so a eventemitter that wraps a try catch) takes 1ms self time, that's a lot of time
1. Task.prototype.exit takes 1ms self time, that's a lot of time
1. Cycling through the nested events to listen for them takes 1ms of self time, that is a lot of time

**Comment by balupton** [comment]
Chrome's TaskGroup (ES6 babel version) profile (8 seconds):

https://www.dropbox.com/s/alsdwj0u97sm09c/taskgroup-es6-babel.cpuprofile?dl=0

**Comment by balupton** [comment]
Okay I was able to get it down to 3.5 seconds by removing maps and disabling nested events by default.

Removing the next tick stuff actually makes it go way slower, as the stack continues to grow.

@pflannery any idea how to evaluate the heap snapshots to tell if it has cleaned up properly, I have no clue how I am meant to read those snapshots.

**Comment by pflannery** [comment]
Firstly what I do is replace babel's _applyConstructor code (at the very top of TaskGroup output) so we can see the Task class in the heap list, otherwise they get created as Object

```js
var _applyConstructor = function (Constructor, args) {
  return new (Function.prototype.bind.apply(Constructor, [null].concat(args)));
};
```

The way I evaluate the heaps is to take three snapshots, one at the start, one just after run and then one when it's completed - [like this example](https://gist.github.com/pflannery/11bfe55a14dd54c768f7)
Then I use the comparison view between them which shows what's been added or removed. 

Also If you take a snapshot when destroy is emitted then that will show you what's left in memory after clean up.

**Comment by balupton** [comment]
> The three snapshot technique will go away soon as there is now a Record Heap Allocations in chrome which gives us a timeline of heap allocations but it's not yet working in the current v8-profiler so we cant do it yet :'(

Working on Chrome Canary with https://github.com/bevry/taskgroup/issues/19#issuecomment-85403753

Will try your suggestion, hopefully it will make the snapshots make sense.

**Comment by pflannery** [comment]
just seen that using async WriteFile [here](https://github.com/bevry/taskgroup/blob/es6/es6/test/util.js#L21) was keeping all the snapshots and strings in memory so each snapshot ended up containing the previous snapshots. I've created PR #20 to resolve this

**Comment by balupton** [comment]
> just seen that using async WriteFile here was keeping all the snapshots and strings in memory so each snapshot ended up containing the previous snapshots. I've created PR #20 to resolve this

Merged, but still getting segfaults on heaps (neither sync or async work). I don't get segfaults on profile writing sync or async (both work).

**Comment by pflannery** [comment]
> Merged, but still getting segfaults on heaps (neither sync or async work). I don't get segfaults on profile writing sync or async (both work).
For me it does grow large but doesn't get a seg fault with either async or sync..

The latest changes seems to of cleared this problem. I wonder how much of this was removing domains and how much was removing coffee?

**Comment by balupton** [comment]
@pflannery domains are still enabled by default in node land (just disabled in browser land)

**Comment by balupton** [comment]
@pflannery does the latest commits make this any better? Notably: https://github.com/bevry/taskgroup/commit/9edcd0f91641403a64ee87dc26fc9fa6438fa567#commitcomment-10406680

**Comment by pflannery** [comment]
@balupton seem my [reply](https://github.com/bevry/taskgroup/commit/9edcd0f91641403a64ee87dc26fc9fa6438fa567#commitcomment-10413551) to you commit comments

**Comment by balupton** [comment]
Closed with v5.0.0 - https://github.com/bevry/taskgroup/blob/master/HISTORY.md#v500-2016-june-4
